components/ws-manager-mk2/service/manager.go

```diff
@@ -296,8 +304,87 @@ func (m *WorkspaceManagerServer) Subscribe(req *api.SubscribeRequest, srv api.Wo
-	return nil, status.Errorf(codes.Unimplemented, "method MarkActive not implemented")
+// MarkActive records a workspace as being active which prevents it from timing out
+func (wsm *WorkspaceManagerServer) MarkActive(ctx context.Context, req *wsmanapi.MarkActiveRequest) (res *wsmanapi.MarkActiveResponse, err error) {
+	//nolint:ineffassign
+	span, ctx := tracing.FromContext(ctx, "MarkActive")
+	tracing.ApplyOWI(span, log.OWI("", "", req.Id))
+	defer tracing.FinishSpan(span, &err)
+
+	workspaceID := req.Id
+
+	var ws workspacev1.Workspace
+	err = wsm.Client.Get(ctx, types.NamespacedName{Namespace: wsm.Config.Namespace, Name: req.Id}, &ws)
+	if errors.IsNotFound(err) {
+		return nil, status.Errorf(codes.NotFound, "workspace %s does not exist", req.Id)
+	}
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "cannot mark workspace: %v", err)
+	}
+
+	var firstUserActivity *timestamppb.Timestamp
+	for _, c := range ws.Status.Conditions {
+		if c.Type == string(workspacev1.WorkspaceConditionFirstUserActivity) {
+			firstUserActivity = timestamppb.New(c.LastTransitionTime.Time)
+		}
+	}
+
+	// if user already mark workspace as active and this request has IgnoreIfActive flag, just simple ignore it
+	if firstUserActivity != nil && req.IgnoreIfActive {
+		return &api.MarkActiveResponse{}, nil
+	}
+
+	// We do not keep the last activity in the workspace resource to limit the load we're placing
+	// on the K8S master in check. Thus, this state lives locally in a map.
+	now := time.Now().UTC()
+	wsm.activity.Store(req.Id, now)
+
+	// We do however maintain the the "closed" flag as annotation on the workspace. This flag should not change
+	// very often and provides a better UX if it persists across ws-manager restarts.
+	isMarkedClosed := conditionPresentAndTrue(ws.Status.Conditions, string(workspacev1.WorkspaceConditionClosed))
+	if req.Closed && !isMarkedClosed {
+		err = wsm.modifyWorkspace(ctx, req.Id, true, func(ws *workspacev1.Workspace) error {
+			ws.Status.Conditions = addUniqueCondition(ws.Status.Conditions, metav1.Condition{
+				Type:               string(workspacev1.WorkspaceConditionClosed),
+				Status:             metav1.ConditionTrue,
+				LastTransitionTime: metav1.NewTime(now),
+				Reason:             "MarkActiveRequest",
+			})
+			return nil
+		})
+	} else if !req.Closed && isMarkedClosed {
+		err = wsm.modifyWorkspace(ctx, req.Id, true, func(ws *workspacev1.Workspace) error {
+			ws.Status.Conditions = addUniqueCondition(ws.Status.Conditions, metav1.Condition{
+				Type:               string(workspacev1.WorkspaceConditionClosed),
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: metav1.NewTime(now),
+				Reason:             "MarkActiveRequest",
+			})
+			return nil
+		})
+	}
+	if err != nil {
+		log.WithError(err).WithFields(log.OWI("", "", workspaceID)).Warn("was unable to mark workspace properly")
```

Comment:
Suggestion, non-blocking:
Include `req.Closed` and `isMarkedClosed` in the warning message, so that we know which path and error were encountered.